### PR TITLE
Match branch names by pattern – EXP-575

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description

This add a new setting to enable prebuilds on branches matched by glob patterns, e.g. `feature-*, main, bug-*`

Screenshot 2023-09-19 at 10 41 38

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 5f91d66</samp>

This pull request adds a new field `prebuild_branch_pattern` to the project settings API and the corresponding types in Go, TypeScript, and protobuf. This field allows users to specify a branch pattern for triggering prebuilds on their projects.

</details>

## Related Issue(s)

Fixes EXP-575

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-match-branches</li>
	<li><b>🔗 URL</b> - <a href="https://at-match-branches.preview.gitpod-dev.com/workspaces" target="_blank">at-match-branches.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - at-match-branches-gha.17118</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-at-match-branches%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/prebuilds/prebuild-manager.ts [hunk 1]
```diff
@@ -379,7 +380,34 @@ export class PrebuildManager {
...
+            const branchNamePattern = project.settings?.prebuildBranchPattern?.trim();
+            if (!branchNamePattern) {
+                // no pattern provided is treated as run on all branches
+                return { shouldRun: true, reason: "all-branches-selected" };
+            }
+
+            for (let pattern of branchNamePattern.split(",")) {
+                // prepending `**/` as branch names can be 'refs/heads/something/feature-x'
+                // and we want to allow simple patterns like: `feature-*`
+                pattern = "**/" + pattern.trim();
```
Comment (svenefftinge) [inline components/server/src/prebuilds/prebuild-manager.ts:398]: 👍

**Review by svenefftinge — APPROVED** [review]
LGTM

**Review by gtsiolis — COMMENTED** [review]
Adding a first round of UX comments. 🥊 

Looks shippable. :shipit:

**Pull request merged**